lib/internal/webstreams/adapters.js

```diff
@@ -0,0 +1,746 @@
+
+const { UV_EOF } = internalBinding('uv');
+
+function newWritableStreamFromStreamWritable(streamWritable) {
+  if (typeof streamWritable?._writableState !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'streamWritable',
+      'stream.Writable',
+      streamWritable);
+  }
+
+  const highWaterMark = streamWritable.writableHighWaterMark;
+  const strategy =
+    streamWritable.writableObjectMode ?
+      new CountQueuingStrategy({ highWaterMark }) :
+      new ByteLengthQueuingStrategy({ highWaterMark });
+
+  let controller;
+  let backpressurePromise;
+  let stream;
+  let closed = false;
+  let errored = false;
+
+  function onClose() {
+    closed = true;
+    streamWritable.off('close', onClose);
+    streamWritable.off('drain', onDrain);
+    streamWritable.off('error', onError);
+    setPromiseHandled(stream.close());
+  }
+
+  function onDrain() {
+    if (backpressurePromise !== undefined)
+      backpressurePromise.resolve();
+  }
+
+  function onError(error) {
+    errored = true;
+    if (backpressurePromise !== undefined)
+      backpressurePromise.reject(error);
+    controller.error(error);
+  }
+
+  streamWritable.once('close', onClose);
+  streamWritable.once('error', onError);
+  streamWritable.on('drain', onDrain);
+
+  stream = new WritableStream({
+    start(c) { controller = c; },
+
+    async write(chunk) {
+      if (!streamWritable.write(chunk)) {
+        backpressurePromise = createDeferredPromise();
+        return PromisePrototypeFinally(
+          backpressurePromise.promise, () => {
+            backpressurePromise = undefined;
+          });
+      }
+    },
+
+    abort(reason) {
+      if (!streamWritable.destroyed && !errored)
+        streamWritable.destroy(reason);
+    },
+
+    close() {
+      if (!streamWritable.writableEnded && !closed)
+        streamWritable.end();
+    },
+  }, strategy);
+
+  return stream;
+}
+
+function newReadableStreamFromStreamReadable(streamReadable) {
+  if (typeof streamReadable?._readableState !== 'object') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'streamReadable',
+      'stream.Readable',
+      streamReadable);
+  }
+
+  const objectMode = streamReadable.readableObjectMode;
+  const highWaterMark = streamReadable.readableHighWaterMark;
+  const strategy =
+    objectMode ?
+      new CountQueuingStrategy({ highWaterMark }) :
+      { highWaterMark };
+
+  const type = !objectMode ? 'bytes' : undefined;
+  const autoAllocateChunkSize = !objectMode ? highWaterMark : undefined;
+
+  let controller;
+  let closed = false;
+
+  function onClose() {
+    closed = true;
+    streamReadable.off('close', onClose);
+    streamReadable.off('data', onData);
+    streamReadable.off('end', onEnd);
+    streamReadable.off('error', onError);
+  }
+
+  function onError(error) {
+    controller.error(error);
+  }
+
+  function onData(chunk) {
+    // Copy the Buffer to detach it from the pool.
+    if (Buffer.isBuffer(chunk))
+      chunk = new Uint8Array(chunk);
```

Comment:
Shouldn't these only be copied if `objectMode == false`? Or can you also have a `Readable` with `objectMode = true` that contains `Buffer` chunks from a pool? 🤔